### PR TITLE
Refactor to a Lerna Monorepo

### LINKED ISSUE #94: Refactor to a Lerna Monorepo
### ✅ Prerequisites

- [x] Did you perform a cursory search of open issues? Is this feature already requested elsewhere?
- [x] Are you reporting to the correct repository (`magic-sdk`)?

### ✨ Feature Request

Soon we will be offering open-source Magic SDK extensions. To support these upcoming new libraries more easily, we should refactor to a monorepo using [Lerna](https://github.com/lerna/lerna).

**EDIT:**

Another great reason for this feature is the blockers encountered during the Redwood JS integration. See the [open PR](https://github.com/magiclabs/magic-js/pull/101) for a more detailed explanation.

### PR DESCRIPTION
**Pull request by smithki**
### 📦 Pull Request

Following the Magic SDK integration into [Redwood JS](https://github.com/redwoodjs/redwood), an issue in our approach to multi-platform support (Web + React Native) became abundantly clear: TypeScript is highly limited in its flexibility with regards to global type definitions. Basically, TypeScript expects each NPM module to contain one, environment-specific package.

If a developer imports any NPM package that contains a reference to, or dependency upon some global types, those types are pulled in—regardless of the project’s requirements. This was causing a bug where React Native types would conflict with DOM types even if only the Web bundle is imported.

This issue was not commonly encountered because developers are largely making use of Webpack to bundle Magic SDK with their applications. The Redwood JS integration was the first time a developer attempted to wrap our package within another NPM module, leading to a TypeScript compiler error. At the moment, the issue is mitigated by disabling React Native typings within Magic SDK, but this is unsatisfactory in the long-term. A preferable approach is to distribute React Native-compatible code as a separate NPM package. This is a breaking change, but I think it’s worth doing to prevent future headaches.

The following packages would be published under this new monorepo architecture:

- `@magic-sdk/provider@1.0.0`
- `@magic-sdk/types@1.0.0`
- `@magic-sdk/react-native@1.0.0`
- `magic-sdk@2.0.0`

### 🗜 Versioning

(Check _one!_)

- [ ] Patch: Bug Fix?
- [ ] Minor: New Feature?
- [x] Major: Breaking Change?

### ✅ Fixed Issues

- Fixes #94 

### 🚨 Test instructions

`yarn build` then `yarn test`

### ⚠️ Update `CHANGELOG.md`

- [x] I have updated the `Upcoming Changes` section of `CHANGELOG.md` with context related to this Pull Request.

### CODE REVIEW

File: packages/react-native/src/index.ts [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+  /* istanbul ignore next */
+  global.atob = b64Encoded => {
+    return Buffer.from(b64Encoded, 'base64').toString('binary');
+  };
+}
+
+/* istanbul ignore next */
+// We cannot test this code path without causing problems for Ava & TS Node.
+// Unfortunatley, we must ignore it.
+if (typeof URL === 'undefined') {
```
Comment (Ethella) [inline packages/react-native/src/index.ts:41]: I get URL “Not implemented” error while testing with the canary version. https://github.com/facebook/react-native/issues/25187
 
I think it's due to the backward compatibility has been removed here. https://github.com/magiclabs/magic-js/blob/monorepo/packages/provider/src/util/url.ts

Comment (Ethella) [reply packages/react-native/src/index.ts:41]: You can test it on their snack without further configuration on your local

https://snack.expo.io/@jerry-liu/magic-link-test

Comment (smithki) [reply packages/react-native/src/index.ts:41]: I’ll take a look today, see if I can narrow down the cause

Comment (smithki) [reply packages/react-native/src/index.ts:41]: I think I've been able to resolve this in `@magic-sdk/react-native@1.0.0-canary.1`.

**Comment by smithki** [comment]
:rocket: PR was released in [`@magic-sdk/provider@3.0.1`](https://github.com/magiclabs/magic-js/releases/tag/%40magic-sdk/provider%403.0.1), [`@magic-sdk/react-native@3.0.1`](https://github.com/magiclabs/magic-js/releases/tag/%40magic-sdk/react-native%403.0.1), [`@magic-sdk/types@2.0.1`](https://github.com/magiclabs/magic-js/releases/tag/%40magic-sdk/types%402.0.1), [`magic-sdk@3.0.1`](https://github.com/magiclabs/magic-js/releases/tag/magic-sdk%403.0.1) :rocket:

**Pull request merged**